### PR TITLE
DEV-203: Add ability for .slp files to define shader variants

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.atlassian.net/browse/DEV-203

Adds support for defines in slp files, like such:
`DEFINES (normalmap translucent:f unlit:f/lightmap:f)/shadow fade:f/forward:f deformed:v/deformeddq:v`

Features:
- builds all combinations of vertex shaders, fragment shaders, and programs using all the space-separated defines
- defines can be grouped with parentheses
- `/` specifies "either/or" defines for things that are mutually exclusive
- `:v` specifies a define is only use in the vertex shader, `:f` specifies it's only used in the fragment shader, so that we aren't wasting time producing unused variants.

For example, the slp above will define the following vertex enums:
```
model,
model_normalmap,
model_shadow,
model_deformed,
model_normalmap_deformed,
model_shadow_deformed,
model_deformeddq,
model_normalmap_deformeddq,
model_shadow_deformeddq,
```

The following fragment shaders:
```
model,
model_normalmap,
model_translucent,
model_normalmap_translucent,
model_unlit,
model_normalmap_unlit,
model_translucent_unlit,
model_normalmap_translucent_unlit,
model_lightmap,
model_normalmap_lightmap,
model_translucent_lightmap,
model_normalmap_translucent_lightmap,
model_shadow,
model_fade,
model_normalmap_fade,
model_translucent_fade,
model_normalmap_translucent_fade,
model_unlit_fade,
model_normalmap_unlit_fade,
model_translucent_unlit_fade,
model_normalmap_translucent_unlit_fade,
model_lightmap_fade,
model_normalmap_lightmap_fade,
model_translucent_lightmap_fade,
model_normalmap_translucent_lightmap_fade,
model_shadow_fade,
model_forward,
model_normalmap_forward,
model_translucent_forward,
model_normalmap_translucent_forward,
model_unlit_forward,
model_normalmap_unlit_forward,
model_translucent_unlit_forward,
model_normalmap_translucent_unlit_forward,
model_lightmap_forward,
model_normalmap_lightmap_forward,
model_translucent_lightmap_forward,
model_normalmap_translucent_lightmap_forward,
model_shadow_forward,
```

And the following programs:
```
model = (vertex::model << 16) | fragment::model,
model_normalmap = (vertex::model_normalmap << 16) | fragment::model_normalmap,
model_translucent = (vertex::model << 16) | fragment::model_translucent,
model_normalmap_translucent = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent,
model_unlit = (vertex::model << 16) | fragment::model_unlit,
model_normalmap_unlit = (vertex::model_normalmap << 16) | fragment::model_normalmap_unlit,
model_translucent_unlit = (vertex::model << 16) | fragment::model_translucent_unlit,
model_normalmap_translucent_unlit = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_unlit,
model_lightmap = (vertex::model << 16) | fragment::model_lightmap,
model_normalmap_lightmap = (vertex::model_normalmap << 16) | fragment::model_normalmap_lightmap,
model_translucent_lightmap = (vertex::model << 16) | fragment::model_translucent_lightmap,
model_normalmap_translucent_lightmap = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_lightmap,
model_shadow = (vertex::model_shadow << 16) | fragment::model_shadow,
model_fade = (vertex::model << 16) | fragment::model_fade,
model_normalmap_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_fade,
model_translucent_fade = (vertex::model << 16) | fragment::model_translucent_fade,
model_normalmap_translucent_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_fade,
model_unlit_fade = (vertex::model << 16) | fragment::model_unlit_fade,
model_normalmap_unlit_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_unlit_fade,
model_translucent_unlit_fade = (vertex::model << 16) | fragment::model_translucent_unlit_fade,
model_normalmap_translucent_unlit_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_unlit_fade,
model_lightmap_fade = (vertex::model << 16) | fragment::model_lightmap_fade,
model_normalmap_lightmap_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_lightmap_fade,
model_translucent_lightmap_fade = (vertex::model << 16) | fragment::model_translucent_lightmap_fade,
model_normalmap_translucent_lightmap_fade = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_lightmap_fade,
model_shadow_fade = (vertex::model_shadow << 16) | fragment::model_shadow_fade,
model_forward = (vertex::model << 16) | fragment::model_forward,
model_normalmap_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_forward,
model_translucent_forward = (vertex::model << 16) | fragment::model_translucent_forward,
model_normalmap_translucent_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_forward,
model_unlit_forward = (vertex::model << 16) | fragment::model_unlit_forward,
model_normalmap_unlit_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_unlit_forward,
model_translucent_unlit_forward = (vertex::model << 16) | fragment::model_translucent_unlit_forward,
model_normalmap_translucent_unlit_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_unlit_forward,
model_lightmap_forward = (vertex::model << 16) | fragment::model_lightmap_forward,
model_normalmap_lightmap_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_lightmap_forward,
model_translucent_lightmap_forward = (vertex::model << 16) | fragment::model_translucent_lightmap_forward,
model_normalmap_translucent_lightmap_forward = (vertex::model_normalmap << 16) | fragment::model_normalmap_translucent_lightmap_forward,
model_shadow_forward = (vertex::model_shadow << 16) | fragment::model_shadow_forward,
model_deformed = (vertex::model_deformed << 16) | fragment::model,
model_normalmap_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap,
model_translucent_deformed = (vertex::model_deformed << 16) | fragment::model_translucent,
model_normalmap_translucent_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent,
model_unlit_deformed = (vertex::model_deformed << 16) | fragment::model_unlit,
model_normalmap_unlit_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_unlit,
model_translucent_unlit_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_unlit,
model_normalmap_translucent_unlit_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_unlit,
model_lightmap_deformed = (vertex::model_deformed << 16) | fragment::model_lightmap,
model_normalmap_lightmap_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_lightmap,
model_translucent_lightmap_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_lightmap,
model_normalmap_translucent_lightmap_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_lightmap,
model_shadow_deformed = (vertex::model_shadow_deformed << 16) | fragment::model_shadow,
model_fade_deformed = (vertex::model_deformed << 16) | fragment::model_fade,
model_normalmap_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_fade,
model_translucent_fade_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_fade,
model_normalmap_translucent_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_fade,
model_unlit_fade_deformed = (vertex::model_deformed << 16) | fragment::model_unlit_fade,
model_normalmap_unlit_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_unlit_fade,
model_translucent_unlit_fade_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_unlit_fade,
model_normalmap_translucent_unlit_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_unlit_fade,
model_lightmap_fade_deformed = (vertex::model_deformed << 16) | fragment::model_lightmap_fade,
model_normalmap_lightmap_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_lightmap_fade,
model_translucent_lightmap_fade_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_lightmap_fade,
model_normalmap_translucent_lightmap_fade_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_lightmap_fade,
model_shadow_fade_deformed = (vertex::model_shadow_deformed << 16) | fragment::model_shadow_fade,
model_forward_deformed = (vertex::model_deformed << 16) | fragment::model_forward,
model_normalmap_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_forward,
model_translucent_forward_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_forward,
model_normalmap_translucent_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_forward,
model_unlit_forward_deformed = (vertex::model_deformed << 16) | fragment::model_unlit_forward,
model_normalmap_unlit_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_unlit_forward,
model_translucent_unlit_forward_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_unlit_forward,
model_normalmap_translucent_unlit_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_unlit_forward,
model_lightmap_forward_deformed = (vertex::model_deformed << 16) | fragment::model_lightmap_forward,
model_normalmap_lightmap_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_lightmap_forward,
model_translucent_lightmap_forward_deformed = (vertex::model_deformed << 16) | fragment::model_translucent_lightmap_forward,
model_normalmap_translucent_lightmap_forward_deformed = (vertex::model_normalmap_deformed << 16) | fragment::model_normalmap_translucent_lightmap_forward,
model_shadow_forward_deformed = (vertex::model_shadow_deformed << 16) | fragment::model_shadow_forward,
model_deformeddq = (vertex::model_deformeddq << 16) | fragment::model,
model_normalmap_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap,
model_translucent_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent,
model_normalmap_translucent_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent,
model_unlit_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_unlit,
model_normalmap_unlit_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_unlit,
model_translucent_unlit_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_unlit,
model_normalmap_translucent_unlit_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_unlit,
model_lightmap_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_lightmap,
model_normalmap_lightmap_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_lightmap,
model_translucent_lightmap_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_lightmap,
model_normalmap_translucent_lightmap_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_lightmap,
model_shadow_deformeddq = (vertex::model_shadow_deformeddq << 16) | fragment::model_shadow,
model_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_fade,
model_normalmap_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_fade,
model_translucent_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_fade,
model_normalmap_translucent_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_fade,
model_unlit_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_unlit_fade,
model_normalmap_unlit_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_unlit_fade,
model_translucent_unlit_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_unlit_fade,
model_normalmap_translucent_unlit_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_unlit_fade,
model_lightmap_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_lightmap_fade,
model_normalmap_lightmap_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_lightmap_fade,
model_translucent_lightmap_fade_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_lightmap_fade,
model_normalmap_translucent_lightmap_fade_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_lightmap_fade,
model_shadow_fade_deformeddq = (vertex::model_shadow_deformeddq << 16) | fragment::model_shadow_fade,
model_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_forward,
model_normalmap_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_forward,
model_translucent_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_forward,
model_normalmap_translucent_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_forward,
model_unlit_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_unlit_forward,
model_normalmap_unlit_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_unlit_forward,
model_translucent_unlit_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_unlit_forward,
model_normalmap_translucent_unlit_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_unlit_forward,
model_lightmap_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_lightmap_forward,
model_normalmap_lightmap_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_lightmap_forward,
model_translucent_lightmap_forward_deformeddq = (vertex::model_deformeddq << 16) | fragment::model_translucent_lightmap_forward,
model_normalmap_translucent_lightmap_forward_deformeddq = (vertex::model_normalmap_deformeddq << 16) | fragment::model_normalmap_translucent_lightmap_forward,
model_shadow_forward_deformeddq = (vertex::model_shadow_deformeddq << 16) | fragment::model_shadow_forward,
```

### CODE REVIEW

File: cmake/macros/AutoScribeShader.cmake [hunk 1]
```diff
@@ -66,6 +66,9 @@ macro(AUTOSCRIBE_PLATFORM_SHADER)
     list(APPEND SHADER_GEN_LINE ${TEMP_PATH})
     file(RELATIVE_PATH TEMP_PATH ${CMAKE_SOURCE_DIR} ${AUTOSCRIBE_OUTPUT_FILE})
     list(APPEND SHADER_GEN_LINE ${TEMP_PATH})
+    if (NOT("${DEFINES}" STREQUAL ""))
+        list(APPEND SHADER_GEN_LINE "defines:${DEFINES}")
+    endif()
```
Comment (SamGondelman) [inline cmake/macros/AutoScribeShader.cmake]: We add an additional section to shadergen.txt that specifies the defines (this is parsed in shadergen.py)

File: cmake/macros/AutoScribeShader.cmake [hunk 2]
```diff
@@ -135,11 +142,72 @@ macro(AUTOSCRIBE_SHADER)
         endif()
     endif()
 
-    string(CONCAT SHADER_ENUMS "${SHADER_ENUMS}" "${SHADER_NAME} = ${SHADER_COUNT},\n")
+    string(CONCAT SHADER_LIST "${SHADER_LIST}" "${SHADER_NAME} = ${SHADER_COUNT},\n")
     string(CONCAT SHADER_SHADERS_ARRAY  "${SHADER_SHADERS_ARRAY}" "${SHADER_COUNT},\n")
     MATH(EXPR SHADER_COUNT "${SHADER_COUNT}+1")
 endmacro()
 
+function(GENERATE_DEFINES_LIST_HELPER INPUT_LIST RETURN_LIST)
```
Comment (SamGondelman) [inline cmake/macros/AutoScribeShader.cmake]: This function takes in the list of defines (which would look like: `(normalmap;translucent:f;unlit:f/lightmap:f)/shadow;fade:f/forward:f;deformed:v/deformeddq:v`) and handles parentheses and slashes, producing the final list of all combinations, which will look like `normalmap;normalmap_translucent:f;normalmap_unlit:f_deformed:v`, etc.)

File: cmake/macros/AutoScribeShader.cmake [hunk 3]
```diff
@@ -135,11 +142,72 @@ macro(AUTOSCRIBE_SHADER)
...
+    string(LENGTH "${INPUT_LIST}" STR_LENGTH)
+    set(OPEN_INDEX -1)
+    set(STR_INDEX 0)
+    set(NESTED_DEPTH 0)
+    while ("${STR_INDEX}" LESS "${STR_LENGTH}")
```
Comment (SamGondelman) [inline cmake/macros/AutoScribeShader.cmake]: this while loop handles parentheses, looking for matching ( and ) and then calling GENERATE_DEFINES_LIST_HELPER recursively on that.

File: cmake/macros/AutoScribeShader.cmake [hunk 4]
```diff
@@ -216,12 +290,72 @@ macro(AUTOSCRIBE_SHADER_LIB)
...
 
-            set(PROGRAM_ENTRY "${PROGRAM_NAME} = (${AUTOSCRIBE_PROGRAM_VERTEX} << 16) |  ${AUTOSCRIBE_PROGRAM_FRAGMENT},\n")
-            string(CONCAT SHADER_ENUMS "${SHADER_ENUMS}" "${PROGRAM_ENTRY}")
+            string(CONCAT PROGRAM_ENUMS "${PROGRAM_ENUMS}" "${PROGRAM_NAME} = (${AUTOSCRIBE_PROGRAM_VERTEX} << 16) | ${AUTOSCRIBE_PROGRAM_FRAGMENT},\n")
             string(CONCAT SHADER_PROGRAMS_ARRAY "${SHADER_PROGRAMS_ARRAY} ${SHADER_NAMESPACE}::program::${PROGRAM_NAME},\n")
+
+            foreach(DEFINES ${DEFINES_LIST})
+                set(ORIG_DEFINES "${DEFINES}")
+
+                string(REPLACE ":v" "" VERTEX_DEFINES "${ORIG_DEFINES}")
```
Comment (SamGondelman) [inline cmake/macros/AutoScribeShader.cmake]: Below here we handle `:v` and `:f`.  The program name includes both, but the vertex and fragment names remove the elements with `:f` and `:v` respectively, and only have to call AUTOSCRIBE_SHADER if they don't have those.

File: tools/shadergen.py [hunk 5]
```diff
@@ -184,19 +196,23 @@ def processCommand(line):
...
+        print('Processing file {} dialect {} variant {} defines {}'.format(scribeFile, dialect, variant, defines))
         if args.dry_run:
             return True
 
-        scribeDepCache.gen(scribeFile, libs, dialect, variant)
+        scribeDepCache.gen(scribeFile, libs, dialect, variant, defines)
         scribeArgs = getCommonScribeArgs(scribeFile, libs)
         for header in getDialectAndVariantHeaders(dialect, variant, args.extensions):
             scribeArgs.extend(['-H', header])
+        for define in getDefines(defines):
```
Comment (SamGondelman) [inline tools/shadergen.py]: For each define in the define combination for this file, we add a scribe define.  Shaders will be able to add specialized code like this:
```
<@if HIFI_USE_FORWARD@>
...
<@else@>
...
<@endif@>
```

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19732/HighFidelity-Beta-PR15662-a6f2dce.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19732/HighFidelity-Beta-Interface-PR15662-a6f2dce.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19732/HighFidelity-Beta-PR15662-a6f2dce.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19732/HighFidelity-Quest-Beta-PR15662-a6f2dce.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19732/HighFidelity-Beta-PR15662-a6f2dce.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19732/HighFidelity-Beta-Interface-PR15662-a6f2dce.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19745/HighFidelity-Beta-PR15662-68d2f95.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19745/HighFidelity-Beta-Interface-PR15662-68d2f95.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19745/HighFidelity-Beta-PR15662-68d2f95.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19745/HighFidelity-Quest-Beta-PR15662-68d2f95.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19745/HighFidelity-Beta-PR15662-68d2f95.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19745/HighFidelity-Beta-Interface-PR15662-68d2f95.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20233/HighFidelity-Beta-PR15662-9813117.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20233/HighFidelity-Quest-Beta-PR15662-9813117.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20233/HighFidelity-Beta-PR15662-9813117.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20233/HighFidelity-Beta-Interface-PR15662-9813117.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20233/HighFidelity-Beta-PR15662-9813117.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20233/HighFidelity-Beta-Interface-PR15662-9813117.exe

**Pull request merged**